### PR TITLE
Pytorch MatMul Substitution

### PR DESCRIPTION
**Pull request by itai-berman**
## Pull Request Description:
Add substitution to support MatMul function & operator by doing Reshape->Split->2D MatMul->Stack->Reshape

## Checklist before requesting a review:
- [x] I set the appropriate labels on the pull request.
- [x] I have added/updated the release note draft (if necessary).
- [x] I have updated the documentation to reflect my changes (if necessary).
- [x] All function and files are well documented. 
- [x] All function and classes have type hints.
- [x] There is a licenses in all file.
- [x] The function and variable names are informative. 
- [x] I have checked for code duplications.
- [x] I have added new unittest (if necessary).

### CODE REVIEW

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/matmul_decomposition.py [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+            weights={},
+            layer_class=torch.reshape,
+            op_call_args=[params.other_reshape_out_shape],
+            op_call_kwargs={},
+            functional_op=torch.reshape
+        )
+        # Connect reshapes to graph
+        # input_in_edge, other_in_edge = graph.in_edges(matmul_node)
+        graph.add_node(input_reshape_node)
+        # graph.add_edge(input_in_edge[0], input_reshape_node, **graph.get_edge_data(*input_in_edge, 0))
```
Comment (elad-c) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/matmul_decomposition.py:229]: remove commented operations.

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/matmul_decomposition.py [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+        # graph.add_edge(other_in_edge[0], other_reshape_node, **graph.get_edge_data(*other_in_edge, 0))
+
+        return input_reshape_node, other_reshape_node
+
+    @staticmethod
+    def _unbind_reshaped(graph: Graph,
+                        matmul_node: BaseNode,
+                        input_reshape_node: BaseNode,
+                        other_reshape_node: BaseNode,
+                        params: MatMulParams) -> List[BaseNode]:
```
Comment (elad-c) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/matmul_decomposition.py:240]: func node. see other functions that need to fix this.

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/matmul_decomposition.py [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+            weights={},
+            layer_class=operator.getitem,
+            op_call_args=[dim_index],
+            op_call_kwargs={},
+            functional_op=operator.getitem
+        )
+        graph.add_node_with_in_edges(get_input_node, [input_unbind_node], [dim_index])
+
+        # Get the other in index dim_index
+        get_other_node = FunctionalNode(
```
Comment (elad-c) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/matmul_decomposition.py:321]: Why do you need the get_item node?

File: tests/pytorch_tests/model_tests/feature_models/matmul_test.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+class MatMulFNet(torch.nn.Module):
+    """
+    Model for testing MatMul function
+    """
+    def __init__(self):
+        super(MatMulFNet, self).__init__()
+
+    def forward(self, x):
+        x_t = torch.transpose(x, -1, -2)
```
Comment (elad-c) [inline tests/pytorch_tests/model_tests/feature_models/matmul_test.py:31]: This doesn't test inputs with different number of dimensions.
For example: need to test `(B, d1, d2, d3, a, b) @ (B, a, b)`, `(B, a, b) @ (B, d1, d2, a, b)`

File: tests/pytorch_tests/model_tests/feature_models/matmul_test.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    """
+    Model for testing MatMul operator
+    """
+    def forward(self, x):
+        x_t = torch.transpose(x, -1, -2)
+        x = x @ x_t
+        return x
+
+
+class MatMulNetBaseTest(BasePytorchTest):
```
Comment (elad-c) [inline tests/pytorch_tests/model_tests/feature_models/matmul_test.py:45]: For all tests: use only float TPC

**Review by elad-c — CHANGES_REQUESTED** [review]

**Review by elad-c — APPROVED** [review]
